std/traits.d

```diff
@@ -3897,6 +3898,58 @@ enum hasMember(T, string name) = __traits(hasMember, T, name);
     static assert(hasMember!(S, "foo"));
 }
 
+/**
+   True if `T1` and `T2` may be incomparable, in the way that any
+   comparison involving `float.nan` will yield false.
+
+   This is true for built-in floating-point types, false for all other
+   built-in types.
+
+   User-defined types may have the same behavior by providing an
+   implementation of `opCmp` that returns float.nan when the compared
+   values are incomparable.
+ */
+template hasNanComparison(T1, T2 = T1)
+{
+    static if (isFloatingPoint!T1)
+        enum hasNanComparison = true;
+    else static if (!isAggregateType!T1)
+        enum hasNanComparison = false;
+    else static if (!is(typeof((T1 a, T2 b) => a.opCmp(b))))
+        enum hasNanComparison = false;
+    else
+        enum hasNanComparison = hasNanComparison!(ReturnType!((T1 a, T2 b) => a.opCmp(b)));
```

Comment:
What is the reason for the recursion instead of `isFloatingPoint!(ReturnType!((T1 a, T2 b) => a.opCmp(b)))`? If there's some subtlety here, please add a comment explaining it.

Reply:
`opCmp` could return a type that isn't floating-point but has an `opCmp` returning float. This tower of `opCmp` could be multiple layers deep. I don't actually see a good reason to do this, but that doesn't mean a good reason doesn't exist.